userspace/tcpvsadm.8
.\"
.\"     tcpvsadm(8) manual page
.\"
.\"	$Id: tcpvsadm.8,v 1.2.2.3 2004/12/08 14:06:31 wensong Exp $
.\"
.\"     Authors: Wensong Zhang <wensong@linux-vs.org>
.\"
.\"     Changes:
.\"
.\"     This program is free software; you can redistribute it and/or modify
.\"     it under the terms of the GNU General Public License as published by
.\"     the Free Software Foundation; either version 2 of the License, or
.\"     (at your option) any later version.
.\"
.\"     This program is distributed in the hope that it will be useful,
.\"     but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"     GNU General Public License for more details.
.\"
.\"     You should have received a copy of the GNU General Public License
.\"     along with this program; if not, write to the Free Software
.\"     Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.\"
.TH TCPVSADM 8 "31 October 2004" "LVS Administration" "Linux Administrator's Guide"
.UC 4
.SH NAME
tcpvsadm \- Kernel TCP Virtual Server administration
.SH SYNOPSIS
.B tcpvsadm -A|E -i \fIident\fP [-s \fIscheduler\fP] [-P \fIport\fP]
.br
.B tcpvsadm -D -i \fIident\fP
.br
.B tcpvsadm -F
.br
.B tcpvsadm -a|e -i \fIident\fP -r \fIserver-address\fP [-w \fIweight\fP]
.br
.B tcpvsadm -d -i \fIident\fP -r \fIserver-address\fP
.br
.B tcpvsadm --add-rule -i \fIident\fP -p \fIpattern\fP -r \fIserver-address\fP [-m \fImatchnum\fP]
.br
.B tcpvsadm --del-rule -i \fIident\fP -p \fIpattern\fP -r \fIserver-address\fP
.br
.B tcpvsadm -L [options]
.br
.B tcpvsadm -f \fIconfig-file\fP
.br
.B tcpvsadm --start|stop [-i \fIident\fP]
.br
.B tcpvsadm -h
.SH DESCRIPTION
\fBTcpvsadm\fR is used to set up, maintain or inspect the TCP Virtual
Server table inside the Linux kernel (KTCPVS). KTCPVS is a Layer-7
switching software in the Linux Virtual Server framework, and can do
application-level load balancing (such as content-based scheduling)
inside the Linux kernel. The reason of implementing it inside the
kernel is to avoid the overhead of context switching and memory
copying between user-space and kernel-space. Although the scalability
of KTCPVS is lower than that of IPVS (IP Virtual Server), it is
flexible, because the content of request is known before the request
is redirected to real server.
.SS COMMANDS
\fBtcpvsadm\fR(8) recognises the commands described below.
.TP
.B -A, --add-service
Add a TCP virtual service.
.TP
.B -E, --edit-service
Edit a TCP virtual service.
.TP
.B -D, --delete-service
Delete a TCP virtual service, along with any associated real servers
and rules.
.TP
.B -F, --flush
Flush the TCP virtual server table.
.TP
.B -a, --add-server
Add a real server to a TCP virtual service.
.TP
.B -e, --edit-server
Edit a real server in a TCP virtual service.
.TP
.B -d, --delete-server
Remove a real server from a TCP virtual service.
.TP
.B --add-rule
Add a content-based scheduling rule to a TCP virtual
service. Currently, the rule is simple, if a request content matchs a
\fIpattern\fP, then the request will be redirected to the specified
server.
.TP
.B --del-rule
Remove a content-based scheduling rule from a TCP virtual service.
.TP
.B -L, --list
List the TCP virtual server table if no argument is specified. If a
service \fIident\fP is selected, list this service only. The exact
output is affected by the other arguments given.
.TP
.B -f --load-configfile \fIconfig-file\fP
Load the TCP virtual server table from the specified
\fIconfig-file\fP.  The original table will be flushed before loading
from the \fIconfig-file\fP.
.TP
.B --start [-i \fIident\fP]
Start all the services in the TCP virtual server table if no argument
is specified. If a service \fIident\fP is selected, start this service
only.
.TP
.B --stop [-i \fIident\fP]
Stop all the services in the TCP virtual server table if no argument
is specified. If a service \fIident\fP is selected, stop this service
only.
.TP
.B -h, --help
Display a description of the command syntax.
.SS PARAMETERS
The commands above accept or require zero or more of the following
parameters.
.TP
.B -i, --ident \fIident\fP
The \fIident\fP is a string of alphabetic and numeric characters, to
identify the name of a TCP virtual service.
.TP
.B -s, --scheduler \fIscheduling-method\fP
The \fIscheduling-method\fP is an algorithm for allocating TCP
connections to real servers. The scheduling algorithms are implemented
as kernel modules. Two are shipped with KTCPVS now:
.sp
\fBwlc\fR - Weighted Least-Connection: assign more jobs to servers
with fewer jobs and relative to the real servers' weight.
.sp
\fBhttp\fR - HTTP content-based scheduling: assign jobs to servers
according to the specified content-based scheduling rules.
.sp
\fBhhttp\fR - Hash-based HTTP scheduling: assign jobs to servers
according to the specified content-based scheduling rules where
regular expression matches are hashed and the virtual server is
selected based on the resulting hash value.
.sp
\fBNote:\fR If a scheduling module cannot be automatically loaded by
modprobe(8), the module may be manually inserted into the kernel using
insmod(8).
.TP
.B -l, --listen \fIserver-address[:port]\fP
Specify the ip address (and port number) that a TCP virtual service
listens to.  The \fIserver-address\fP is the \fIhost\fP address of a
real server, and possibly plus \fIport\fP. \fIHost\fP can be either a
plain IP address or a hostname. \fIPort\fP can be either a plain port
number or the service name of port.
.TP
.B -P, --port \fIport\fP
Specify the port number that a TCP virtual service listens to.
.TP
.B -r, --real-server \fIserver-address\fP
Real server that an associated request for service may be assigned to.
The \fIserver-address\fP is the \fIhost\fP address of a real server,
and may plus \fIport\fP. \fIHost\fP can be either a plain IP address
or a hostname. \fIPort\fP can be either a plain port number or the
service name of port. \fIport\fP may be ommitted, in which case  the
destination port on the real server will be the destination port of
the request sent to the virtual service.
.TP
.B -w, --weight \fIweight\fP
\fIWeight\fP is an integer specifying the capacity  of a server
relative to the others in the pool. The valid values of \fIweight\fP
are 0 through to 65535. The default is 1. Quiescent servers are
specified with a weight of zero. A quiescent server will receive no
new jobs but still serve the existing jobs, for all scheduling
algorithms distributed with the Linux Virtual Server. Setting a
quiescent server may be useful if the server is overloaded or needs to
be taken out of service for maintenance.
.TP
.B -p, --pattern \fIpattern\fP
\fIPattern\fP is a string specifying a matching pattern, and it is
interpreted as an extended regular expression. The pattern ".*" or
"^/.*" is used to specify the default server. When no other server
pattern is matched, the default server will be used.
.TP
.B -n, --numeric
Numeric output.  IP addresses and port numbers will be printed in
numeric format rather than as as host names and services respectively,
which is the  default.
.SH EXAMPLES
The following commands configure a TCP virtual service to distribute
incoming requests addressed to port 80 at local address to three web
servers, according the content-based scheduling rules.
.PP
.nf
tcpvsadm -A -i web -s http -P 80
tcpvsadm -a -i web -r 192.168.10.1:80
tcpvsadm -a -i web -r 192.168.10.2:80
tcpvsadm -a -i web -r 192.168.10.3:80
tcpvsadm --add-rule -i web --pattern=^/cgi-bin/ -r 192.168.10.1:80
tcpvsadm --add-rule -i web --pattern=^/images/ -r 192.168.10.2:80
tcpvsadm --add-rule -i web --pattern=.* -r 192.168.10.3:80
.fi
.PP
Alternatively, this could be achieved by "tcpvsadm -f config-file",
the config file is as follows:
.PP
.nf
virtual web {
    listen = 0.0.0.0:80
    scheduler = http
    startservers = 8
    maxclients = 256
    minspareservers = 6
    maxspareservers = 18
    server = 192.168.10.3:80 1
    server = 192.168.10.2:80 1
    server = 192.168.10.1:80 1
    rule = pattern "^/cgi-bin/"  use server 192.168.10.1:80

ChangeLog
Kernel TCP Virtual Server - Version 0.0.18 - December 18, 2004

2004-12-14  Alexandre Cassen  <acassen@linux-vs.org>

	* tcp_vs.c (tcp_vs_relay_socket): fixed a lock issue at inbound
	while performing socket relay.  This issue only appear on SMP box.


Kernel TCP Virtual Server - Version 0.0.17 - December 08, 2004

2004-12-07  Alexandre Cassen  <acassen@linux-vs.org>

	* Added timeout support to connection handler.
	  Added new sysctl var 'read_timeout' to set this new stuff. The
	  default value is 180 seconds. This new option provides stalled
	  connections hunting.

2004-11-30  Wensong Zhang  <wensong@linux-vs.org>

	* userspace/tcpvsadm.c (main): fixed the bug of the --listen
	option.


Kernel TCP Virtual Server - Version 0.0.16 - October 31, 2004

2004-10-31  Wensong Zhang  <wensong@linux-vs.org>

	* integrate the hhttp scheduler patch from Philipp Klaus
	<pklaus@futurelab.ch>, which is to send requests to server based
	on the hash value of matched request.

	* userspace/tcpvsadm.c: added the --listen option to specify the
	ip address (and port number) that a TCP virtual service listens to.

	* EXAMPLES: added.


Kernel TCP Virtual Server - Version 0.0.15 - July 8, 2003

2003-07-08  Wensong Zhang  <wensong@linux-vs.org>

	* tcp_vs_phttp.c: a little bit code tidy up.

2003-06-15  Wensong Zhang  <wensong@linux-vs.org>

	* tcp_vs_http_trans.h: splitted from tcp_vs.h.

	* tcp_vs_http_parser.h: splitted from tcp_vs.h.


Kernel TCP Virtual Server - Version 0.0.14 - June 14, 2003

2003-06-14  Wensong Zhang  <wensong@linux-vs.org>

	* Merged the patch from Hai Long <david_lung@yahoo.com>

	split the code for HTTP transport from tcp_vs_http_parser.c into
	tcp_vs_http_trans.c.

	Some tunings in the tcp_vs_chttp.c and tcp_vs_phttp.c.

	* Some code tidy up.

2003-06-07  Wensong Zhang  <wensong@linux-vs.org>

	* Merged the patch from Hai Long <david_lung@yahoo.com>

	A large part of the HTTP parsing engine has been rewritten.

	some code tidy up in the tcp_vs_phttp.c and tcp_vs_chttp.c.

2003-05-28  Wensong Zhang  <wensong@linux-vs.org>

	* tcp_vs_chttp.c: fixed the wrong way of calling spin_lock and
	spin_unlock.


Kernel TCP Virtual Server - Version 0.0.13 - May 28, 2003

2003-05-28  Wensong Zhang  <wensong@linux-vs.org>

	* Merged the minor patch from Hai Long <david_lung@yahoo.com> to
	fix up the function strdup.

2003-05-23  Wensong Zhang  <wensong@linux-vs.org>

	* tcp_vs_ctl.c (tcp_vs_edit_service, tcp_vs_add_service): fixed to
	return -ENOENT when the scheduler is not found.

	* Merged the patch from Hai Long <david_lung@yahoo.com>

	Added a new scheduler ktcpvs_chttp to do cookie-injection load
	balancing for web servers.

	Split the code into a stand-alone HTTP parsing engine.

	Included the AVL code from GNU libavl and adapted it to use inside
	the kernel.

2003-05-21  Wensong Zhang  <wensong@linux-vs.org>

	* tcp_vs_timer.c (tcp_vs_slowtimer_init): removed the buggy line of
	assigning slowtimer_jiffies with the system jiffies.

2003-04-27  Wensong Zhang  <wensong@linux-vs.org>

	* some code tidy up.


Kernel TCP Virtual Server - Version 0.0.12 - April 7, 2003

2003-04-07  Wensong Zhang  <wensong@linux-vs.org>

	* tcp_vs_phttp.c (http_read_line): merged the patch from Hai Long
	<david_lung@yahoo.com>, to remove the max_line_size limit in the
	persistent http scheduler.

2003-04-05  Wensong Zhang  <wensong@linux-vs.org>

	* tcp_vs_phttp.c (relay_multiparts): added the missing continue
	statement, and changed the schedule_timeout(HZ) to schedule(),
	because the thread only needs to wait for socket data.

2003-02-27  Wensong Zhang  <wensong@linux-vs.org>

	* userspace/tcpvsadm.c (modprobe_ktcpvs): added to load the ktcpvs
	module automatically by modprobe.

2003-02-25  Wensong Zhang  <wensong@linux-vs.org>

	* userspace/Makefile (MANDIR): applied the patch from Alejandro
	Mery Pellegrini <amery@innernet.cl> to add the $(MANDIR) variable,
	so that it can be overloaded by the generic build-system of
	rocklinux.


userspace/tcpvsadm.c
/*
 *      tcpvsadm - TCP Virtual Server ADMinistration program
 *
 *      Version: $Id: tcpvsadm.c,v 1.12.2.4 2004/12/08 14:06:31 wensong Exp $
 *
 *      Authors: Wensong Zhang <wensong@linuxvirtualserver.org>
 *
 *      Note that a lot code is taken from ipvsadm.c.
 *
 *      This program is free software; you can redistribute it and/or modify
 *      it under the terms of the GNU General Public License as published by
 *      the Free Software Foundation; either version 2 of the License, or
 *      (at your option) any later version.
 *
 *      This program is distributed in the hope that it will be useful,
 *      but WITHOUT ANY WARRANTY; without even the implied warranty of
 *      MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *      GNU General Public License for more details.
 *
 *      You should have received a copy of the GNU General Public License
 *      along with this program; if not, write to the Free Software
 *      Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#undef __KERNEL__		/* Makefile lazyness ;) */
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <getopt.h>
#include <netdb.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <ctype.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <sys/param.h>
#include <arpa/inet.h>

#include <net/if.h>
#include <netinet/ip_icmp.h>
#include <netinet/udp.h>
#include <netinet/tcp.h>

#include "tcp_vs.h"
#include "helper.h"
#include "tcpvs_config.h"
#include "libtcpvs/libtcpvs.h"

static const char *program;
#define TCPVSADM_VERSION_DATE	"21-May-2007"
static const char *program_version = VERSION ", " TCPVSADM_VERSION_DATE;

/* default scheduler */
#define DEFAULT_SCHEDULER	"http"

/* check the options based on the commands_v_options table */
static void generic_opt_check(int command, int options);
static void set_command(unsigned int *cmd, unsigned int newcmd);
static void set_option(unsigned int *options, unsigned int option);

static void tryhelp_exit(const int status);
static void usage_exit(const int status);
static void fail(int err, char *msg, ...);

static int list_service(struct tcp_vs_ident *id, unsigned int format);
static int list_all(unsigned int format);
static int load_configfile(char *cf);
static int modprobe_ktcpvs(void);


#define CMD_NONE		0x0000U
#define CMD_ADD			0x0001U
#define CMD_EDIT		0x0002U
#define CMD_DEL			0x0004U
#define CMD_FLUSH		0x0008U
#define CMD_LIST		0x0010U
#define CMD_ADDDEST		0x0020U
#define CMD_EDITDEST		0x0040U
#define CMD_DELDEST		0x0080U
#define CMD_ADDRULE		0x0100U
#define CMD_DELRULE		0x0200U
#define CMD_START		0x0400U
#define CMD_STOP		0x0800U
#define CMD_LOADCF		0x1000U
#define NUMBER_OF_CMD		13

static const char *cmdnames[] = {
	"add-service",
	"edit-service",
	"delele-service",
	"flush",
	"list",
	"add-server",
	"delete-server",
	"edit-server",
	"add-rule",
	"del-rule",
	"start",
	"stop",
	"load-configfile",
};

#define OPT_NONE	0x00000
#define OPT_NUMERIC	0x00001
#define OPT_IDENT	0x00002
#define OPT_SCHEDULER	0x00004
#define OPT_SERVERADDR	0x00008
#define OPT_SERVERPORT	0x00010
#define OPT_REALSERVER	0x00020
#define OPT_WEIGHT	0x00040
#define OPT_PATTERN	0x00080
#define OPT_MATCHNUM	0x00100
#define OPT_LISTEN	0x00200
#define NUMBER_OF_OPT	11

static const char *optnames[] = {
	"numeric",
	"ident",
	"scheduler",
	"serveraddr",
	"serverport",
	"real-server",
	"weight",
	"pattern",
	"match",
	"listen",
};

/*
 * Table of legal combinations of commands and options.
 * Key:
 *  '+'  compulsory
 *  'x'  illegal
 *  '1'  exclusive (only one '1' option can be supplied)
 *  ' '  optional
 */
static const char commands_v_options[NUMBER_OF_CMD][NUMBER_OF_OPT] = {
/*             -n   -i   -s   ads  prt  -r   -w   -p   -m   -l */
/*ADD*/       {'x', '+', ' ', ' ', ' ', 'x', 'x', 'x', 'x', ' '},
/*EDIT*/      {'x', '+', ' ', ' ', ' ', 'x', 'x', 'x', 'x', 'x'},
/*DEL*/       {'x', '+', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x'},
/*FLUSH*/     {'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x'},
/*LIST*/      {' ', ' ', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x'},
/*ADD-SERVER*/{'x', '+', 'x', 'x', 'x', '+', ' ', 'x', 'x', 'x'},
/*DEL-SERVER*/{'x', '+', 'x', 'x', 'x', '+', 'x', 'x', 'x', 'x'},
/*EDIT-SRV*/  {'x', '+', 'x', 'x', 'x', '+', ' ', 'x', 'x', 'x'},
/*ADD-RULE*/  {'x', '+', 'x', 'x', 'x', '+', 'x', '+', ' ', 'x'},
/*DEL-RULE*/  {'x', '+', 'x', 'x', 'x', '+', 'x', '+', 'x', 'x'},
/*START*/     {'x', '1', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x'},
/*STOP*/      {'x', '1', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x'},
/*LOAD-CF*/   {'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x'},
};

static struct option long_options[] = {
	{"add-service", 0, 0, 'A'},
	{"edit-service", 0, 0, 'E'},
	{"delete-service", 0, 0, 'D'},
	{"flush", 0, 0, 'F'},
	{"list", 0, 0, 'L'},
	{"add-server", 0, 0, 'a'},
	{"edit-server", 0, 0, 'e'},
	{"delete-server", 0, 0, 'd'},
	{"add-rule", 0, 0, '1'},
	{"del-rule", 0, 0, '2'},
	{"start", 0, 0, '3'},
	{"stop", 0, 0, '4'},
	{"help", 0, 0, 'h'},
	{"ident", 1, 0, 'i'},
	{"listen", 0, 0, 'l'},
	{"port", 1, 0, 'P'},
	{"scheduler", 1, 0, 's'},
	{"real-server", 1, 0, 'r'},
	{"weight", 1, 0, 'w'},
	{"pattern", 1, 0, 'p'},
	{"match", 1, 0, 'm'},
	{"numeric", 0, 0, 'n'},
	{"load-configfile", 1, 0, 'f'},
	{0, 0, 0, 0}
};


int
main(int argc, char **argv)
{
	const char *optstring = "AEDFaedLf:hi:s:l:P:r:p:m:n";
	int c, parse;
	char cf[128];
	unsigned int command = CMD_NONE;
	unsigned int options = OPT_NONE;
	unsigned int format = FMT_NONE;
	struct tcp_vs_ident ident;
	struct tcp_vs_config conf;
	struct tcp_vs_dest_u dest;
	struct tcp_vs_rule_u rule;
	int result = 0;

	if (tcpvs_init()) {
		/* try to insmod the ip_vs module if ipvs_init failed */
		if (modprobe_ktcpvs() || tcpvs_init())
			fail(2, "%s", tcpvs_strerror(errno));
	}

	/* If no other arguement, list the ktcpvs service table */
	if (argc == 1) {
		list_all(format);
		tcpvs_close();
		return 0;
	}

	program = argv[0];
	memset(&ident, 0, sizeof(ident));
	memset(&conf, 0, sizeof(conf));
	memset(&dest, 0, sizeof(dest));
	memset(&rule, 0, sizeof(rule));

	/* default values */
	conf.port = htons(8080);
	conf.maxSpareServers = 18;
	conf.minSpareServers = 6;
	conf.startservers = 8;
	conf.maxClients = 256;
	dest.weight = 1;
	strcpy(conf.sched_name, DEFAULT_SCHEDULER);

	/*
	 *      Parse options
	 */
	if ((c = getopt_long(argc, argv, optstring,
			     long_options, NULL)) == EOF)
		tryhelp_exit(-1);

	switch (c) {
	case 'A':
		set_command(&command, CMD_ADD);
		break;
	case 'E':
		set_command(&command, CMD_EDIT);
		break;
	case 'D':
		set_command(&command, CMD_DEL);
		break;
	case 'a':
		set_command(&command, CMD_ADDDEST);
		break;
	case 'e':
		set_command(&command, CMD_EDITDEST);
		break;
	case 'd':
		set_command(&command, CMD_DELDEST);
		break;
	case 'F':
		set_command(&command, CMD_FLUSH);
		break;
	case '1':
		set_command(&command, CMD_ADDRULE);
		break;
	case '2':
		set_command(&command, CMD_DELRULE);
		break;
	case '3':
		set_command(&command, CMD_START);
		break;
	case '4':
		set_command(&command, CMD_STOP);
		break;
	case 'L':
		set_command(&command, CMD_LIST);
		break;
	case 'f':
		set_command(&command, CMD_LOADCF);
		strncpy(cf, optarg, 128);
		break;
	case 'h':
		usage_exit(0);
		break;
	default:
		tryhelp_exit(-1);
	}

	while ((c = getopt_long(argc, argv, optstring,
				long_options, NULL)) != EOF) {
		switch (c) {
		case 'i':
			set_option(&options, OPT_IDENT);
			strncpy(ident.name, optarg,
				KTCPVS_IDENTNAME_MAXLEN);
			break;
		case 's':
			set_option(&options, OPT_SCHEDULER);
			strncpy(conf.sched_name, optarg,
				KTCPVS_SCHEDNAME_MAXLEN);
			break;
		case 'P':
			set_option(&options, OPT_SERVERPORT);
			parse = string_to_number(optarg, 1, 65534);
			if (parse == -1)
				fail(2, "illegal server port specified");
			conf.port = htons(parse);
			break;
		case 'l':
			set_option(&options, OPT_LISTEN);
			parse = parse_addrport(optarg, IPPROTO_TCP,
					       &conf.addr, &conf.port);;
			if (parse == 0)
				fail(2, "illegal address:port specified");
			break;
		case 'r':
			set_option(&options, OPT_REALSERVER);
			parse = parse_addrport(optarg, IPPROTO_TCP,
					       &dest.addr, &dest.port);
			if (parse == 0)
				fail(2, "illegal address:port specified");
			break;
		case 'w':
			set_option(&options, OPT_WEIGHT);
			if ((dest.weight =
			     string_to_number(optarg, 0, 65535)) == -1)
				fail(2, "illegal weight specified");
			break;
		case 'p':
			set_option(&options, OPT_PATTERN);
			strncpy(rule.pattern, optarg,
				KTCPVS_PATTERN_MAXLEN);
			rule.len = strlen(optarg);
			break;
		case 'm':
			set_option(&options, OPT_MATCHNUM);
			parse = string_to_number(optarg, 0, 10);
			if (parse == -1)
				fail(2, "illegal match number specified");
			rule.match_num = parse;
			break;
		case 'n':
			set_option(&options, OPT_NUMERIC);
			format |= FMT_NUMERIC;
			break;
		default:
			fail(2, "invalid option");
		}
	}

	if (optind < argc)
		fail(2, "unknown arguments found in command line");

	generic_opt_check(command, options);

	if (command == CMD_ADDRULE || command == CMD_DELRULE) {
		rule.addr = dest.addr;
		rule.port = dest.port;
	}

	switch (command) {
	case CMD_ADD:
		result = tcpvs_add_service(&ident, &conf);
		break;
	case CMD_EDIT:
		result = tcpvs_edit_service(&ident, &conf);
		break;
	case CMD_DEL:
		result = tcpvs_del_service(&ident);
		break;
	case CMD_FLUSH:
		result = tcpvs_flush();
		break;
	case CMD_LIST:
		if (options & OPT_IDENT)
			result = list_service(&ident, format);
		else
			result = list_all(format);
		break;
	case CMD_ADDDEST:
		result = tcpvs_add_dest(&ident, &dest);
		break;
	case CMD_EDITDEST:
		result = tcpvs_edit_dest(&ident, &dest);
		break;
	case CMD_DELDEST:
		result = tcpvs_del_dest(&ident, &dest);
		break;
	case CMD_ADDRULE:
		result = tcpvs_add_rule(&ident, &rule);
		break;
	case CMD_DELRULE:
		result = tcpvs_del_rule(&ident, &rule);
		break;
	case CMD_START:
		result = tcpvs_start_service(&ident);
	case CMD_STOP:
		result = tcpvs_stop_service(&ident);
		break;
	case CMD_LOADCF:
		result = load_configfile(cf);
		break;
	}

	if (result)
		fprintf(stderr, "%s\n", tcpvs_strerror(errno));

	tcpvs_close();
	return 0;
}


static int
modprobe_ktcpvs(void)
{
	char *argv[] =
	    { "/sbin/modprobe", "-s", "-k", "--", "ktcpvs", NULL };
	int child;
	int status;
	int rc;

	if (!(child = fork())) {
		execv(argv[0], argv);
		exit(1);
	}

	rc = waitpid(child, &status, 0);

	if (!WIFEXITED(status) || WEXITSTATUS(status)) {
		return 1;
	}

	return 0;
}


static int
load_configfile(char *cf)
{
	struct tcpvs_config config;
	struct tcpvs_service *svc;
	int i, j;
	int rc;

	if ((rc = tcpvs_flush()))
		return rc;

	rc = tcpvs_parse_config(cf, &config);
	if (rc)
		return rc;

	for (i = 0; i < config.num_services; i++) {
		svc = &config.services[i];
		rc = tcpvs_add_service(&svc->ident, &svc->conf);
		if (rc)
			return rc;

		/* add the destination servers */
		for (j = 0; j < svc->num_dests; j++) {
			rc = tcpvs_add_dest(&svc->ident, &svc->dests[j]);
			if (rc)
				return rc;
		}

		/* add the rules */
		for (j = 0; j < svc->num_rules; j++) {
			rc = tcpvs_add_rule(&svc->ident, &svc->rules[j]);
			if (rc)
				return rc;
		}
	}
	return 0;
}


static void
print_service(struct tcp_vs_service_u *svc, unsigned int format)
{
	struct tcp_vs_get_dests *d;
	struct tcp_vs_get_rules *r;
	struct in_addr laddr;
	char *listen;
	struct in_addr daddr;
	char *dname;
	int i;

	if (!(d = tcpvs_get_dests(svc)))
		exit(1);
	if (!(r = tcpvs_get_rules(svc)))
		exit(1);

	printf("Virtual %s {\n", svc->ident.name);

	laddr.s_addr = svc->conf.addr;
	if (!(listen = addrport_to_anyname(&laddr, ntohs(svc->conf.port),
					   IPPROTO_TCP, format)))
		fail(2, "addrport_to_anyname: %s", strerror(errno));

	printf("    listen = %s\n", listen);
	printf("    scheduler = %s\n", svc->conf.sched_name);
	printf("    startservers = %d\n", svc->conf.startservers);
	printf("    maxclients = %d\n", svc->conf.maxClients);
	printf("    minspareservers = %d\n", svc->conf.minSpareServers);
	printf("    maxspareservers = %d\n", svc->conf.maxSpareServers);

	/* print the redirect address */
	if (svc->conf.redirect_port) {
		struct in_addr addr;
		char *name;

		addr.s_addr = svc->conf.redirect_addr;
		name = addrport_to_anyname(&addr,
					   ntohs(svc->conf.redirect_port),
					   IPPROTO_TCP, format);
		if (!name)
			fail(2, "addrport_to_anyname: %s",
			     strerror(errno));
		printf("    redirect = %s\n", name);
		free(name);
	}

	/* print all the destination entries */
	for (i = 0; i < d->num_dests; i++) {
		struct tcp_vs_dest_u *e = &d->entrytable[i];

		daddr.s_addr = e->addr;
		if (!(dname = addrport_to_anyname(&daddr, ntohs(e->port),
						  IPPROTO_TCP, format)))
			fail(2, "addrport_to_anyname: %s",
			     strerror(errno));
		printf("    server = %s %d\n", dname, e->weight);
		free(dname);
	}

	/* print all the rule entries */
	for (i = 0; i < r->num_rules; i++) {
		struct tcp_vs_rule_u *e = &r->entrytable[i];

		daddr.s_addr = e->addr;
		if (!(dname = addrport_to_anyname(&daddr, ntohs(e->port),
						  IPPROTO_TCP, format)))
			fail(2, "addrport_to_anyname: %s",
			     strerror(errno));
		printf("    rule = pattern \"%s\" ",
		       e->pattern);
		if (!strcasecmp(svc->conf.sched_name, "hhttp"))
			printf("match %d ", e->match_num);
		printf("use server %s\n", dname);
		free(dname);
	}

	printf("}\n");
	free(listen);
	free(d);
}


static int
list_service(struct tcp_vs_ident *id, unsigned int format)
{
	struct tcp_vs_service_u *svc;

	if (!(svc = tcpvs_get_service(id)))
		return -1;

	print_service(svc, format);
	free(svc);
	return 0;
}


static int
list_all(unsigned int format)
{
	struct tcp_vs_get_services *get;
	int i;

	printf("TCP Virtual Server version %d.%d.%d\n",
	       NVERSION(tcpvs_info.version));

	if (!(get = tcpvs_get_services()))
		return -1;

	for (i = 0; i < get->num_services; i++)
		print_service(&get->entrytable[i], format);
	free(get);
	return 0;
}


static void
generic_opt_check(int command, int options)
{
	int i, j;
	int last = 0, count = 0;

	/* Check that commands are valid with options. */
	for (i = 0; i < NUMBER_OF_CMD; i++) {
		if (command & (1 << i))
			break;
	}

	for (j = 0; j < NUMBER_OF_OPT; j++) {
		if (!(options & (1 << j))) {
			if (commands_v_options[i][j] == '+')
				fail(2, "You need to supply the '%s' "
				     "option for the '%s' command",
				     optnames[j], cmdnames[i]);
		} else {
			if (commands_v_options[i][j] == 'x')
				fail(2, "Illegal '%s' option with "
				     "the '%s' command",
				     optnames[j], cmdnames[i]);
			if (commands_v_options[i][j] == '1') {
				count++;
				if (count == 1) {
					last = j;
					continue;
				}
				fail(2,
				     "The option '%s' conflicts with the "
				     "'%s' option in the '%s' command",
				     optnames[j], optnames[last],
				     cmdnames[i]);
			}
		}
	}
}

static inline const char *
opt2name(int option)
{
	const char **ptr;
	for (ptr = optnames; option > 1; option >>= 1, ptr++);

	return *ptr;
}

static void
set_command(unsigned int *cmd, unsigned int newcmd)
{
	if (*cmd != CMD_NONE)
		fail(2, "multiple commands specified");
	*cmd = newcmd;
}

static void
set_option(unsigned int *options, unsigned int option)
{
	if (*options & option)
		fail(2, "multiple '%s' options specified",
		     opt2name(option));
	*options |= option;
}


static void
tryhelp_exit(const int status)
{
	fprintf(stderr,
		"Try `%s -h' or '%s --help' for more information.\n",
		program, program);
	exit(status);
}


static void
usage_exit(const int status)
{
	FILE *stream;

	if (status != 0)
		stream = stderr;
	else
		stream = stdout;

	fprintf(stream,
		"%s %s\n"
		"Usage:\n"
		"  %s -A|E -i ident [-s scheduler] [-P port] [-l address[:port]]\n"
		"  %s -D -i ident\n"
		"  %s -F\n"
		"  %s -a|e -i ident -r server-address [-w weight]\n"
		"  %s -d -i ident -r server-address\n"
		"  %s --add-rule -i ident -p pattern -r server-address [-m match-num]\n"
		"  %s --del-rule -i ident -p pattern -r server-address\n"
		"  %s -L [-n]\n"
		"  %s -f config-file\n"
		"  %s --start|stop [-i ident]\n"
		"  %s -h\n\n",
		program, program_version, program, program, program,
		program, program, program, program, program, program,
		program, program);

	fprintf(stream,
		"Commands:\n"
		"Either long or short options are allowed.\n"
		"  --add-service     -A        add virtual service with options\n"
		"  --edit-service    -E        edit virtual service with options\n"
		"  --delete-service  -D        delete virtual service\n"
		"  --flush           -F        flush the whole table\n"
		"  --add-server      -a        add real server with options\n"
		"  --edit-server     -e        edit real server with options\n"
		"  --delete-server   -d        delete real server\n"
		"  --add-rule                  add rule into virtual service\n"
		"  --del-rule                  del rule into virtual service\n"
		"  --list            -L        list the table\n"
		"  --load-configfile -f        load a config file\n"
		"  --start                     start the virtual services\n"
		"  --stop                      stop the virtual services\n"
		"  --help            -h        display this help message\n\n");

	fprintf(stream,
		"Options:\n"
		"  --ident        -i identity          service identity\n"
		"  --scheduler    -s scheduler         one of %s\n"
		"                                      the default scheduler is %s.\n"
		"  --port         -p port              service port number\n"
		"  --match        -m match-num         match number to hash (hhttp only)\n"
		"  --real-server  -r server-address    server-address is host (and port)\n"
		"  --listen       -l server-address    server-address is host (and port)\n"
		"  --weight       -w weight            capacity of real server\n"
		"  --numeric      -n                   numeric output of addresses and ports\n",
		SCHEDULERS, DEFAULT_SCHEDULER);

	exit(status);
}


static void
fail(int err, char *msg, ...)
{
	va_list args;

	va_start(args, msg);
	vfprintf(stderr, msg, args);
	va_end(args);
	fprintf(stderr, "\n");
	exit(err);
}
